centchain/api.go

```diff
@@ -1,74 +1,163 @@
+	return dsa.sapi.RPC.State.GetRuntimeVersionLatest()
+}
+
+func (dsa *defaultSubstrateAPI) GetClient() client.Client {
+	return dsa.sapi.Client
+}
+
+func (dsa *defaultSubstrateAPI) GetStorageLatest(key types.StorageKey, target interface{}) error {
+	return dsa.sapi.RPC.State.GetStorageLatest(key, target)
 }
 
 type api struct {
-	getBlockHash            func(uint64) (types.Hash, error)
-	getRuntimeVersionLatest func() (*types.RuntimeVersion, error)
-	getStorageLatest        func(key types.StorageKey, target interface{}) error
-	getClient               func() client.Client
-	getBlockLatest          func() (*types.SignedBlock, error)
-	getMetadataLatest       func() (*types.Metadata, error)
-	mu                      sync.Mutex
+	sapi     SubstrateAPI
+	config   Config
+	queueSrv *queue.Server
+	accounts map[string]uint32
+	accMu    sync.Mutex // accMu to protect accounts
+	mu       sync.Mutex
 }
 
 // NewAPI returns a new centrifuge chain api.
-func NewAPI(sapi *gsrpc.SubstrateAPI) API {
-	return api{
-		getBlockHash:            sapi.RPC.Chain.GetBlockHash,
-		getRuntimeVersionLatest: sapi.RPC.State.GetRuntimeVersionLatest,
-		getStorageLatest:        sapi.RPC.State.GetStorageLatest,
-		getClient:               func() client.Client { return sapi.Client },
-		getBlockLatest:          sapi.RPC.Chain.GetBlockLatest,
-		getMetadataLatest:       sapi.RPC.State.GetMetadataLatest,
+func NewAPI(sapi SubstrateAPI, config Config, queueSrv *queue.Server) API {
+	return &api{
+		sapi:     sapi,
+		config:   config,
+		queueSrv: queueSrv,
+		accounts: map[string]uint32{},
+		accMu:    sync.Mutex{},
+		mu:       sync.Mutex{},
 	}
 }
 
-func (a api) GetMetadataLatest() (*types.Metadata, error) {
-	return a.getMetadataLatest()
+func (a *api) Call(result interface{}, method string, args ...interface{}) error {
+	return a.sapi.Call(result, method, args...)
 }
 
-func (a api) SubmitExtrinsic(meta *types.Metadata, c types.Call, krp signature.KeyringPair) (txHash types.Hash, bn types.BlockNumber, sig types.Signature, err error) {
-	a.mu.Lock()
-	defer a.mu.Unlock()
+func (a *api) GetMetadataLatest() (*types.Metadata, error) {
+	return a.sapi.GetMetadataLatest()
+}
+
+func (a *api) SubmitExtrinsic(ctx context.Context, meta *types.Metadata, c types.Call, krp signature.KeyringPair) (txHash types.Hash, bn types.BlockNumber, sig types.MultiSignature, err error) {
 	ext := types.NewExtrinsic(c)
 	era := types.ExtrinsicEra{IsMortalEra: false}
 
-	genesisHash, err := a.getBlockHash(0)
-	if err != nil {
-		return txHash, bn, sig, err
-	}
-
-	rv, err := a.getRuntimeVersionLatest()
+	genesisHash, err := a.sapi.GetBlockHash(0)
 	if err != nil {
 		return txHash, bn, sig, err
 	}
 
-	key, err := types.CreateStorageKey(meta, "System", "AccountNonce", krp.PublicKey)
+	rv, err := a.sapi.GetRuntimeVersionLatest()
 	if err != nil {
 		return txHash, bn, sig, err
 	}
 
-	var nonce uint32
-	err = a.getStorageLatest(key, &nonce)
+	nonce, err := contextutil.Nonce(ctx)
```

Comment:
wouldn;t it be better to fetch the latest nonce here instead of fetching from CTX?

Reply:
or maybe a explicit argument to function maybe?

Reply:
I wanted to have the submitExtrinsic as simple as possible and delegate the nonce to the submit with retrial logic, since it is where the latest nonce per account is stored, since now we keep it in memory.

Reply:
I think the CTX is an ok place to put data that is passed downstream.

Reply:
passing nonce explicitly will avoid unneeded type conversions with context and also error checking. At a later time, If we continue to pass nonce down the stream through context, nonce might not be the latest and we may have to deal with txn failure issues. 

For now, this is okay but we need to be careful on this IMO